src/Interpreters/GinFilter.cpp

```diff
@@ -217,6 +212,67 @@ size_t GinFilter::memoryUsageBytes() const
     return memory_usage;
 }
 
+std::vector<uint32_t> GinFilter::getIndices(
+    const GinFilter *filter,
+    const PostingsCacheForStore *cache_store,
+    const MarkRanges &ranges) const
+{
+    chassert(ranges.size() > 0);
+
+    if (filter->getTerms().empty())
+        return {};
+
+    const size_t full_start = ranges.front().begin + 1;
+    const size_t full_end = ranges.back().end + 1;
+
+    GinIndexPostingsList range_bitset;
+    for (const MarkRange &range : ranges)
+        range_bitset.addRange(range.begin + 1, range.end + 2); // Yes, i know how it looks...
+
+    const GinPostingsCachePtr postings_cache = cache_store->getCachedPostings(*filter);
+    GinIndexPostingsList posting_bitset;
+
+    for (const GinSegmentWithRowIdRange &range : rowid_ranges)
+    {
+        if (range.range_end < full_start)
+            continue;
+
+        if (range.range_start > full_end)
+            break;
+
+        GinIndexPostingsList range_matches;
+        range_matches.addRange(range.range_start, range.range_end + 1);
+
+        for (const auto & term_postings : *postings_cache)
+        {
+            /// Check if it is in the same segment by searching for segment_id
+            const GinSegmentedPostingsListContainer & container = term_postings.second;
+            auto container_it = container.find(range.segment_id);
+
+            if (container_it == container.cend()
+                || container_it->second->maximum() < range.range_start
+                || container_it->second->minimum() > range.range_end)
+            {
+                range_matches.removeRange(range.range_start, range.range_end + 1);
```

Comment:
looking at the removeRange from CRoaring, it seems it's bit complex and probably costly. Can we just somehow skip it or initialize ranges_matches empty and use `OR (|)` afterwards for the matching range?